### DISCUSSION TITLE
Move some types & traits to standalone crate?

### DISCUSSION THREAD

**Discussion by Berrysoft** [tokio-rs/tokio-uring#109] [Ideas]
Thanks for your work on bringing io-uring to async Rust! Inspired by this project, I wrote [`tokio-iocp`](https://crates.io/crates/tokio-iocp) to bring Windows IOCP to tokio. Both framework, io-uring and IOCP, are proactors, and the wrapper projects needs many similar types and traits.

For example, the buffers are passed into the read/write methods, which could be abstracted to new traits, maybe `AsyncReadOwned` and `AsyncWriteOwned`. The traits for the buffers, `IoBuf` and `IoBufMut`, together with type `Slice`, could also be shared between the wrappers. The frameworks on different OS differ, but the wrapper API could be the same, and finally we can make our way to unified API, for example, `tokio-proactor`.

I'm asking that if you could split some general types, mainly `IoBuf*`, `Slice`, and if possible, some general IO traits, to a standalone crate. And if you are interested in, you can take a look into `tokio-iocp`, whose API is very much like `tokio-uring`. Anyway, thank you all for your contributions for Rust!

**Comment by Noah-Kennedy** [comment]
I think it would be a good idea to have a separate crate with IoBuf/Slice and owned IO traits. We could add a new crate to the tokio org for this and re-export it through uring.

@carllerche how do you feel about doing this?
